### PR TITLE
Parse and emit years from 1900 to 9999, and remove environment variable dependence on Android

### PR DESCRIPTION
**Pull request by BillyONeal**
Update date formatting and parsing to accept all years between 1900 and 9999.

This unblocks the Azure Storage SDK which uses 9999-12-31 as a sentinel value for "no expiration time" which was broken by the locale insensitive date formatting's dependence on _mkgmtime, with its 1970-3000 range. By morally reimplementing _mkgmtime, this change also lets us remove the Android workaround of changing the TZ environment variable.

* Release\tests\functional\utils\datetime.cpp
  * Turn on tests for values that would need 64 bits represented as a time_t, even for 32 bit time_t platforms.
  * Add year 1900 and 9999 test cases, and remove year 3000+ and 1969- negative tests.

* Release\tests\functional\http\*
  * Drive by fix for new VS2019 warnings about narrowing.

* Release\src\utilities\asyncrt_utils.cpp
  * Remove dependence on _mkgmtime and gmtime_r/gmtime_s by implementing replacements loosely inspired by both the C runtime's date operations and .NET's handling of the Gregorian calendar's 400 year cycle.

### CODE REVIEW

File: Release/tests/functional/http/listener/request_handler_tests.cpp [hunk 1]
```diff
@@ -445,14 +445,12 @@ SUITE(request_handler_tests)
             headers[U("Request")] = U("Upload");
             headers[U("ImgNr")] = U("1");
 
-            utility::char_t* pdata = new utility::char_t[nbytes];
+            char* pdata = new char[nbytes];
```
Comment (BillyONeal) [inline Release/tests/functional/http/listener/request_handler_tests.cpp]: Whoops

File: Release/src/utilities/asyncrt_utils.cpp [hunk 2]
```diff
@@ -756,9 +758,9 @@ utility::string_t datetime::to_string(date_format format) const
         throw std::out_of_range("The requested year exceeds the year 9999.");
     }
 
-    const int64_t epochAdjusted = static_cast<int64_t>(m_interval) - Jan11900InNT;
+    const int64_t epochAdjusted = static_cast<int64_t>(m_interval) - NtTo1900OffsetInterval;
     const int64_t secondsSince1900 = static_cast<int64_t>(epochAdjusted / _secondTicks); // convert to seconds
```
Comment (ras0219-msft) [inline Release/src/utilities/asyncrt_utils.cpp]: Is this static cast required?

File: Release/src/utilities/asyncrt_utils.cpp [hunk 3]
```diff
@@ -991,7 +993,7 @@ datetime __cdecl datetime::from_string(const utility::string_t& dateString, date
 {
     datetime result;
     int64_t secondsSince1900;
```
Comment (ras0219-msft) [inline Release/src/utilities/asyncrt_utils.cpp]: Might be worth initializing to 0

Comment (BillyONeal) [reply Release/src/utilities/asyncrt_utils.cpp]: I explicitly don't want to do that so that if there's any path where it isn't assigned the compiler will warn.

File: Release/src/utilities/asyncrt_utils.cpp (hunk 2 above)
Comment (BillyONeal) [reply Release/src/utilities/asyncrt_utils.cpp]: Yes; interval_type is uint64_t meaning this will warn about signed/unsigned mismatches without the cast. (interval_type should probably be int64_t but I'm scared to change that)

Comment (ras0219-msft) [reply Release/src/utilities/asyncrt_utils.cpp]: I mean the line below -- line 760/762.
```
const int64_t secondsSince1900 = static_cast<int64_t>(epochAdjusted / _secondTicks); // convert to seconds
```
Since `epochAdjusted` is already int64, the static cast should not be needed iiuc

Comment (BillyONeal) [reply Release/src/utilities/asyncrt_utils.cpp]: Hmmmm my 'usual arithmetic conversions paranoia' is showing. I guess that one can probably go now.

Comment (BillyONeal) [reply Release/src/utilities/asyncrt_utils.cpp]: https://github.com/Microsoft/cpprestsdk/commit/b7d3e8b900bc5813deb455c5fff18d9603539265
